pennylane/ops/op_math/controlled_class.py

```diff
@@ -311,18 +337,24 @@ def diagonalizing_gates(self):
         return self.base.diagonalizing_gates()
 
     def decomposition(self):
-        if not all(self.control_values):
-            d = [
-                qml.PauliX(w) for w, val in zip(self.control_wires, self.control_values) if not val
-            ]
-            d += [Controlled(self.base, self.control_wires, work_wires=self.work_wires)]
-            d += [
-                qml.PauliX(w) for w, val in zip(self.control_wires, self.control_values) if not val
-            ]
-
-            return d
-        # More to come.  This will be an extensive PR in and of itself.
-        return super().decomposition()
+        decomp = _decompose_no_control_values(self)
+        if all(self.control_values):
+            if decomp is None:
+                raise qml.operation.DecompositionUndefinedError
+            return decomp
+
+        # We need to add paulis to flip some control wires
+        d = [qml.PauliX(w) for w, val in zip(self.control_wires, self.control_values) if not val]
+
+        if decomp is None:
+            no_control_values = copy(self)
+            no_control_values.hyperparameters["control_values"] = [1] * len(self.control_wires)
+            d.append(no_control_values)
+        else:
+            d += decomp
+
+        d += [qml.PauliX(w) for w, val in zip(self.control_wires, self.control_values) if not val]
```

Comment:
Would it make sense to store the initially created `d` and add a copy here at the end, instead of recreating it? :thinking:

Reply:
`op.expand()` basically creates a tape using `decomposition` as a quantum function.  For this to work correctly, the second set of Pauli's needs to be unique and queued.

Reply:
I see, thanks!